### ISSUE TITLE
Async

### ISSUE THREAD

**Issue by joshnoe** [imaya/zlib.js#49]
Currently, only synchronous compression / decompression is supported. It would be great if a callback could be passed (or better, a promise returned) to provide a way to compress / decompress asynchronously.

**Comment by LuoZijun** [comment]
@joshnoe i think that is one bad idea.

normaly:

`parallel` for `computing` and `asynchronous` for `I/O` .

**Comment by joshnoe** [comment]
You're right, I misspoke. Parallel is what I want.

This is easy to do with a web worker, which is what I ended up doing. It would be nice if it was integrated into the existing API though.

**Comment by LuoZijun** [comment]
@joshnoe  thank you let me know about web woker.
i have do some research about web worker: https://gist.github.com/LuoZijun/9efafd7a69ffb00b0ad8db373073315e

some browser do not allow this:

```javascript
// IE Problem.
var worker = new Worker('data:text/javascript;charset=US-ASCII,onmessage%20%3D%20function%20%28oEvent%29%20%7B%0A%09postMessage%28%7B%0A%09%09%22id%22%3A%20oEvent.data.id%2C%0A%09%09%22evaluated%22%3A%20eval%28oEvent.data.code%29%0A%09%7D%29%3B%0A%7D');
```

i think that is one big problem.

**Comment by joshnoe** [comment]
Good point, but even without support from all browsers, I still think this is a worthwhile feature as long as its browser compatibility is well documented.

You could always fall-back to synchronous if the browser doesn't support web workers.

**Comment by LuoZijun** [comment]
@joshnoe okay, this maybe work.

*zlib.min.js*

```javascript
var Zlib = {
  // include zlib source code
};
// worker code
onmessage = function(evt) {
    var id = evt.data.id;
    var action = evt.data.action;
    var data   = evt.data.data;
    var result = null;
    if ( action == 'compress' ){
        result = (new Zlib.Deflate(data)).compress();
    } else if ( action == 'decompress' ){
        result = (new Zlib.Inflate(data)).decompress();
    }
    postMessage({
        "id"      : request_id,
        "result": result
    });
}
```


*app.js*

```javascript

var worker = new Worker('./zlib.min.js');

// compress data
worker.postMessage({
    "id"    : 1,             // random id or sha1-256 hash
    "action": "compress",
    "data"  : "Hello, 世界！"
});

// decompress data
worker.postMessage({
    "id"    : 1,             // random id or sha1-256 hash
    "action": "decompress",
    "data"  : new Uint8Array([256, 123,97, 96])
});


worker.onmessage = function (evt){
    console.log("Result: ", evt.data);
};
```

**Comment by joshnoe** [comment]
@LuoZijun Yea something like that would work.

**Comment by LuoZijun** [comment]
@joshnoe  :)

**Comment by joshnoe** [comment]
@LuoZijun Are you interested in integrating this into the Library? You could have two new methods: "CompressParallel" and "DecompressParallel". These could return promises which would be resolved when they're complete (when the web worker "onmessage" is called. If the browser doesn't support web workers, it could just run synchronously.

What do you think? I'd be happy to write something up and submit a pull request if you're busy.

**Comment by LuoZijun** [comment]
@joshnoe i'm not this  Library Author, you should talk with @imaya .
at last, i think support parallel compress/decompress is very cool. i like it.
